internal/backend/runtime/omni/controllers/omni/schematic_configuration.go

```diff
@@ -217,3 +216,78 @@ func updateFinalizers(ctx context.Context, r controller.ReaderWriter, extensions
 
 	return r.AddFinalizer(ctx, extensions.Metadata(), SchematicConfigurationControllerName)
 }
+
+type machineExtensions struct {
+	machineStatus     *omni.MachineStatus
+	machineExtensions *omni.MachineExtensions
+	cluster           *omni.Cluster
+	extensionsList    []string
+}
+
+func newMachineExtensions(cluster *omni.Cluster, machineStatus *omni.MachineStatus, extensions *omni.MachineExtensions) (machineExtensions, error) {
+	me := machineExtensions{
+		machineStatus: machineStatus,
+		cluster:       cluster,
+	}
+
+	if extensions != nil && extensions.Metadata().Phase() == resource.PhaseRunning {
+		me.machineExtensions = extensions
+	}
+
+	detected, err := getDetectedExtensions(cluster, machineStatus)
+	if err != nil {
+		return me, err
+	}
+
+	me.extensionsList = append(me.extensionsList, detected...)
+
+	if me.machineExtensions != nil {
+		for _, e := range me.machineExtensions.TypedSpec().Value.Extensions {
+			if slices.Contains(me.extensionsList, e) {
+				continue
+			}
+
+			me.extensionsList = append(me.extensionsList, e)
+		}
+	}
+
+	return me, nil
+}
+
+func (m machineExtensions) shouldGenerateSchematic() bool {
+	// generate schematic for the machine extensions when either machine extensions exists
+	// and contains the explicit empty list for the schematics, or when schematic list is not empty
+	return m.machineExtensions != nil || len(m.extensionsList) != 0
+}
+
+func getDetectedExtensions(cluster *omni.Cluster, machineStatus *omni.MachineStatus) ([]string, error) {
+	if machineStatus.TypedSpec().Value.InitialTalosVersion == "" {
+		return nil, errors.New("machine initial version is not set")
+	}
+
+	initialVersion, err := semver.ParseTolerant(machineStatus.TypedSpec().Value.InitialTalosVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	currentVersion, err := semver.ParseTolerant(cluster.TypedSpec().Value.TalosVersion)
```

Comment:
it's minor right now, but we should do `currentVersion.Pre = nil`, otherwise `v1.6.0-alpha.0 >= v1.6.0` is not true